File: demo/src/main/java/com/example/demo/services/JobQueryService.java
package com.example.demo.services;

import org.springframework.data.domain.Sort;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

@Service
public class JobQueryService {

    public List<Criteria> buildJobKeywordConditions(String keyword) {
        if (keyword == null || keyword.isEmpty()) {
            return new ArrayList<>();
        }
        Pattern searchRegex = Pattern.compile(keyword, Pattern.CASE_INSENSITIVE);
        return List.of(
                Criteria.where("title").regex(searchRegex),
                Criteria.where("description").regex(searchRegex),
                Criteria.where("level").regex(searchRegex),
                Criteria.where("requiredSkills.name").regex(searchRegex)
        );
    }

    public Sort buildSortQuery(String sort) {
        if (sort == null) {
            return Sort.by(Sort.Direction.DESC, "createdAt");
        }
        switch (sort) {
            case "oldest":
                return Sort.by(Sort.Direction.ASC, "createdAt");
            case "newest":
            default:
                return Sort.by(Sort.Direction.DESC, "createdAt");
        }
    }
}

File: backend/src/services/jobQuery.service.js
export const buildJobKeywordConditions = (keyword) => {
  if (!keyword) {
    return [];
  }
  const searchRegex = new RegExp(keyword, 'i');
  return [
    { title: searchRegex },
    { description: searchRegex },
    { level: searchRegex },
    { "requiredSkills.name": searchRegex } // MongoDB tự động quét qua mảng các object
  ];
};


export const buildSortQuery = (sort) => {
  switch (sort) {
    case "oldest":
      return { createdAt: 1 }; // Sort by creation date ascending
    case "newest":
    default:
      return { createdAt: -1 };
  }
};
